src/devices/Sn74hc595/Sn74hc595.cs

```diff
@@ -0,0 +1,361 @@
+        /// <param name="count">Count of (daisy-chained) shift registers. Default/minimum is 1.</param>
+        public Sn74hc595(PinMapping pinMapping, GpioController gpioController = null,  bool shouldDispose = true, int count = 1)
+        {
+            pinMapping.Validate();
+            if (gpioController == null)
+            {
+                gpioController = new GpioController();
+            }
+
+            _controller = gpioController;
+            _shouldDispose = shouldDispose;
+            _pinMapping = pinMapping;
+            _data = _pinMapping.Data;
+            _srclk = _pinMapping.SrClk;
+            _rclk = _pinMapping.RClk;
+            _count = count;
+            SetupPins();
+        }
+
+        /// <summary>
+        /// Initialize a new Sn74hc595 device connected through SPI (uses 3 pins)
+        /// </summary>
+        /// <param name="spiDevice">SpiDevice used for serial communication.</param>
+        /// <param name="count">Count of (daisy-chained) shift registers. Default/minimum is 1.</param>
+        public Sn74hc595(SpiDevice spiDevice, int count = 1)
+        {
+            _spiDevice = spiDevice;
+            _count = count;
+        }
+
+        /// <summary>
+        /// Initialize a new Sn74hc595 device connected through both SPI and GPIO (SPI for writing data; GPIO for configuration; use 4-5 pins)
+        /// </summary>
+        /// <param name="spiDevice">SpiDevice used for serial communication.</param>
+        /// <param name="pinMapping">The pin mapping to use by the binding</param>
+        /// <param name="gpioController">The GPIO Controller used for interrupt handling</param>
+        /// <param name="shouldDispose">True (the default) if the GPIO controller shall be disposed when disposing this instance</param>
+        /// <param name="count">Count of (daisy-chained) shift registers. Default/minimum is 1.</param>
+        public Sn74hc595(SpiDevice spiDevice, PinMapping pinMapping, GpioController gpioController = null, bool shouldDispose = true, int count = 1)
+                  : this(pinMapping, gpioController, shouldDispose, count)
+        {
+            _spiDevice = spiDevice;
+        }
+
+        /// <summary>
+        /// Count of shift units.
+        /// The number of (daisy-chained) shift registers. Minimum is 1.
+        /// Not the count of registers on a single unit.
+        /// </summary>
+        public int Count => _count;
+
+        /// <summary>
+        /// Count of total bits / registers across all (daisy-chained) shift registers.
+        /// Minimum is 8.
+        /// </summary>
+        public int Bits => _count * 8;
+
+        /// <summary>
+        /// Reports if Sn74hc595 is controlled with SPI.
+        /// </summary>
+        public bool UsesSpi => _spiDevice is object;
+
+        /// <summary>
+        /// Reports if Sn74hc595 is controlled with GPIO.
+        /// </summary>
+        public bool UsesGpio => _controller is object;
```

Comment:
Nit: I'm not sure of what a good user scenario would be for these two properties, although I'm not against keeping them.

Reply:
I added those because some of the methods throw in the SPI case.

See:
- https://github.com/dotnet/iot/pull/1104/files#diff-255452453cfa55f55ad064a57e6b07ccR36
- https://github.com/dotnet/iot/pull/1104/files#diff-9de118a60b0d902626974c85247fd348R136

This is the downside of integrating SPI into this API. I think its worth it, but w/o these properties, it is hard to know which APIs are safe to call.